extensions/cornerstone-dicom-sr/src/getSopClassHandlerModule.ts

```diff
@@ -205,69 +210,101 @@ async function _load(displaySet, servicesManager: AppTypes.ServicesManager, exte
     return;
   }
 
-  const { sopClassUids } = newDisplaySet;
+  const { sopClassUids, images } = newDisplaySet;
 
-  // Check if any have the newDisplaySet is the correct SOPClass.
+  /** Check correct SOPClassUID */
   unloadedMeasurements = unloadedMeasurements.filter(measurement =>
-    measurement.coords.some(coord =>
-      sopClassUids.includes(coord.ReferencedSOPSequence.ReferencedSOPClassUID)
-    )
+    measurement.coords.some(coord => {
+      if (coord.ReferencedSOPSequence === undefined && coord.ReferencedFrameOfReferenceSequence) {
+        for (let i = 0; i < images.length; ++i) {
+          const imageMetadata = images[i];
+          if (imageMetadata.FrameOfReferenceUID !== coord.ReferencedFrameOfReferenceSequence) {
+            continue;
+          }
+
+          const sliceNormal = [0, 0, 0];
+          const orientation = imageMetadata.ImageOrientationPatient;
+          sliceNormal[0] = orientation[1] * orientation[5] - orientation[2] * orientation[4];
+          sliceNormal[1] = orientation[2] * orientation[3] - orientation[0] * orientation[5];
+          sliceNormal[2] = orientation[0] * orientation[4] - orientation[1] * orientation[3];
+
+          let distanceAlongNormal = 0;
+          for (let j = 0; j < 3; ++j) {
+            distanceAlongNormal += sliceNormal[j] * imageMetadata.ImagePositionPatient[j];
+          }
+
+          /** Assuming 2 mm tolerance */
+          if (Math.abs(distanceAlongNormal - coord.GraphicData[2]) > 2) {
+            continue;
+          }
+
+          coord.ReferencedSOPSequence = {
+            ReferencedSOPClassUID: imageMetadata.SOPClassUID,
+            ReferencedSOPInstanceUID: imageMetadata.SOPInstanceUID,
+          };
+
+          break;
+        }
+      }
+
+      return (
+        coord.ReferencedSOPSequence &&
+        sopClassUids.includes(coord.ReferencedSOPSequence.ReferencedSOPClassUID)
+      );
+    })
   );
 
   if (unloadedMeasurements.length === 0) {
-    // New displaySet isn't the correct SOPClass, so can't contain the referenced images.
+    /** New displaySet isn't the correct SOPClassso can't contain the referenced images */
     return;
   }
 
   const SOPInstanceUIDs = [];
-
   unloadedMeasurements.forEach(measurement => {
-    const { coords } = measurement;
-
-    coords.forEach(coord => {
-      const SOPInstanceUID = coord.ReferencedSOPSequence.ReferencedSOPInstanceUID;
-
+    measurement.coords.forEach(({ ReferencedSOPSequence }) => {
+      const SOPInstanceUID = ReferencedSOPSequence.ReferencedSOPInstanceUID;
       if (!SOPInstanceUIDs.includes(SOPInstanceUID)) {
         SOPInstanceUIDs.push(SOPInstanceUID);
       }
     });
   });
 
-  const imageIdsForDisplaySet = dataSource.getImageIdsForDisplaySet(newDisplaySet);
-
-  for (const imageId of imageIdsForDisplaySet) {
-    if (!unloadedMeasurements.length) {
-      // All measurements loaded.
-      return;
-    }
-
+  const imageIds = dataSource.getImageIdsForDisplaySet(newDisplaySet);
```

Comment:
If you create a Map<SOP Instance, ImageId[]> up front then, you can iterate over the relevant image ids.  You could even find the exact one of interest by including the frame# in the image id map.  Otherwise very large image id maps will get quite slow, even for a small number of measurements.

Reply:
Updated!